### DISCUSSION TITLE
SBN1661 in C

### DISCUSSION THREAD

**Discussion by Grantiscool** [olikraus/u8g2#1399] [Q&A]
Hi!

I design analogue audio products for my small business and have been working on digital control for future products. I'm fairly new to C, but a couple of weeks after switching from Great Cow BASIC, I've managed to implement most of the control code in my project.

SBN1661-based controllers seem to be ideal for reproducing a frequency response chart, and I've ordered a prototype board using an ATMega164. While the board is fabricated and shipped, I'd love to get an understanding of U8G2 (or even U8Glib, but I realise that's been discontinued).

Call me an idiot, but I've spent two days trying to figure out how to assign pins for parallel* (8080 or 6800) use of the SBN1661. I THINK I've figured this out for U8Glib (via. u8g_init8bit in u8g.h), but I can't figure it out for U8G2. The Wiki seems to explain how to assign pins for Arduino and C++, but not for C. 

I realise there's a lot to learn, but the first step is getting a display working, then I can gradually figure the rest out from there. I'm happy to use the original U8Glib if you guys think it'd be better, but I appreciate that this isn't the correct forum.

Any guidance would be greatly appreciated!

Cheers,

-Grant

(*The final device may use a different protocol, but due to limitations imposed by available thru-hole AVRs and the rest of my circuit, the hardware I2C/SPI pins are unavailable, so it must be parallel for now. I can hack the board and make small changes.)

**Comment by olikraus** [comment]
Part of the answers might be here: https://github.com/olikraus/u8g2/wiki/u8g2as7
In C the display init should look like this:
https://github.com/olikraus/u8g2/blob/3c6460a73f7f310c665cef0af1d2bac49bf6c655/sys/avr/as7/main.c#L174-L176
Obviously you have to modify the setup for your display.

The setup function probably will be u8g2_Setup_sbn1661_122x32_1 or u8g2_Setup_sbn1661_122x32_f (depending on how much memory you can spent for u8g2), see https://github.com/olikraus/u8g2/wiki/u8g2setupc#sbn1661-122x32

Instead of u8x8_byte_4wire_sw_spi you need to use one of the parallel port byte functions:
https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform#communication-callback-eg-u8x8_byte_hw_i2c

And finally you need to provide the gpio and delay callback. You can probably takeover the code from here (including the delay procedures): https://github.com/olikraus/u8g2/blob/3c6460a73f7f310c665cef0af1d2bac49bf6c655/sys/avr/as7/main.c#L112
But you need to add the pins for your parallel interface. This means you need some code, which looks similar to this: 
https://github.com/olikraus/u8g2/blob/3c6460a73f7f310c665cef0af1d2bac49bf6c655/sys/avr/as7/main.c#L131-L136
for U8X8_MSG_GPIO_D0 to U8X8_MSG_GPIO_D7. The full template (including all the case statements) is here: https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform#template-for-the-gpio-and-delay-callback

**Reply by Grantiscool** [reply]
Hi,

I think all the pieces are coming together. Please see the code below, to see if I'm getting the right idea. I was totally ignoring anything that said u8x8! Oops!

The only syntax errors i seem to be getting are path-related, but this is because I haven't set paths yet in MPLab (the menus are different from Atmel Studio, but I'm sure I can figure that out).

There are undoubtedly a lot of device-specific things I need to study in this code, re. timings etc. 

Cheers!


#include <avr/io.h>
#include <u8g2.h>
#include <util/delay.h>

#define F_CPU 8000000UL

#define DISPLAY_DATA_DIR DDRC
#define DISPLAY_DATA_PORT PORTC
#define DISPLAY_DATA_PIN0 0
#define DISPLAY_DATA_PIN1 1
#define DISPLAY_DATA_PIN2 2
#define DISPLAY_DATA_PIN3 3
#define DISPLAY_DATA_PIN4 4
#define DISPLAY_DATA_PIN5 5
#define DISPLAY_DATA_PIN6 6
#define DISPLAY_DATA_PIN7 7 

#define DISPLAY_RW_DIR DDRD 
#define DISPLAY_RW_PORT PORTD
#define DISPLAY_RW_PIN 0

#define DISPLAY_CS1_DIR DDRB
#define DISPLAY_CS1_PORT PORTB
#define DISPLAY_CS1_PIN 1

#define DISPLAY_CS2_DIR DDRB
#define DISPLAY_CS2_PORT PORTB
#define DISPLAY_CS2_PIN 2

#define DISPLAY_DC_DIR DDRB
#define DISPLAY_DC_PORT PORTB
#define DISPLAY_DC_PIN 4

#define DISPLAY_RESET_DIR DDRD
#define DISPLAY_RESET_PORT PORTD
#define DISPLAY_RESET_PIN 3
//Port D because of a slight PCB error! bodge wire time...

#define P_CPU_NS (1000000000UL / F_CPU)

u8g2_t u8g2;

uint8_t u8x8_avr_delay(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
	uint8_t cycles;

	switch(msg)
	{
		case U8X8_MSG_DELAY_NANO:   
			break;
		case U8X8_MSG_DELAY_100NANO:    
            
#define CALL_CYCLES 26UL
#define CALC_CYCLES 4UL
#define RETURN_CYCLES 4UL
#define CYCLES_PER_LOOP 4UL

			cycles = (100UL * arg_int) / (P_CPU_NS * CYCLES_PER_LOOP);

			if(cycles > CALL_CYCLES + RETURN_CYCLES + CALC_CYCLES) 
				break;

			__asm__ __volatile__ (
			"1: sbiw %0,1" "\n\t" // 2 cycles
			"brne 1b" : "=w" (cycles) : "0" (cycles) // 2 cycles
			);
			break;
		case U8X8_MSG_DELAY_10MICRO:    // delay arg_int * 10 micro seconds
			for(int i=0 ; i < arg_int ; i++)
				_delay_us(10);
			break;
		case U8X8_MSG_DELAY_MILLI:      // delay arg_int * 1 milli second
			for(int i=0 ; i < arg_int ; i++)
				_delay_ms(1);
			break;
		default:
			return 0;
	}
	return 1;
}


uint8_t u8x8_avr_gpio_and_delay(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
	switch(msg)
	{
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN0;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN1;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN2;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN3;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN4;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN5;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN6;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN7;
			DISPLAY_RW_DIR |= 1<<DISPLAY_RW_PIN;
			DISPLAY_CS1_DIR |= 1<<DISPLAY_CS1_PIN;
			DISPLAY_CS2_DIR |= 1<<DISPLAY_CS2_PIN;
			DISPLAY_DC_DIR |= 1<<DISPLAY_DC_PIN;
			DISPLAY_RESET_DIR |= 1<<DISPLAY_RESET_PIN;
			break;             
		
		case U8X8_MSG_GPIO_D0:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN0);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN0);
			break;
		case U8X8_MSG_GPIO_D1:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN1);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN1);
			break;
		case U8X8_MSG_GPIO_D2:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN2);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN2);
			break;	
		case U8X8_MSG_GPIO_D3:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN3);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN3);
			break;	
		case U8X8_MSG_GPIO_D4:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN4);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN4);
			break;	
		case U8X8_MSG_GPIO_D5:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN5);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN5);
			break;	
		case U8X8_MSG_GPIO_D6:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN6);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN6);
			break;	
		case U8X8_MSG_GPIO_D7:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN7);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN7);
			break;									
		case U8X8_MSG_GPIO_CS1:        
			if(arg_int)
				DISPLAY_CS1_PORT |= (1<<DISPLAY_CS1_PIN);
			else
				DISPLAY_CS1_PORT &= ~(1<<DISPLAY_CS1_PIN);
			break;
		case U8X8_MSG_GPIO_DC:        
			if(arg_int)
				DISPLAY_DC_PORT |= (1<<DISPLAY_DC_PIN);
			else
				DISPLAY_DC_PORT &= ~(1<<DISPLAY_DC_PIN);
			break;
		case U8X8_MSG_GPIO_RESET:    
			if(arg_int)
				DISPLAY_RESET_PORT |= (1<<DISPLAY_RESET_PIN);
			else
				DISPLAY_RESET_PORT &= ~(1<<DISPLAY_RESET_PIN);
			break;
		case U8X8_MSG_GPIO_E:					if(arg_int)
				DISPLAY_RW_PORT |= (1<<DISPLAY_RW_PIN);
			else
				DISPLAY_RW_PORT &= ~(1<<DISPLAY_RW_PIN);
			break;
		default:
			if (u8x8_avr_delay(u8x8, msg, arg_int, arg_ptr))	// check for any delay msgs
				return 1;
			u8x8_SetGPIOResult(u8x8, 1);      // default return value
			break;
	}
	return 1;
}

int main(void)
{
	u8g2_Setup_sbn1661_122x32_2(u8g2, U8G2_R2, u8x8_byte_sed1520, u8x8_avr_gpio_and_delay);
	u8g2_InitDisplay(&u8g2);
	u8g2_SetPowerSave(&u8g2, 0);
	
	/* full buffer example, setup procedure ends in _f */
	u8g2_ClearBuffer(&u8g2);
	u8g2_SetFont(&u8g2, u8g2_font_ncenB14_tr);
	u8g2_DrawStr(&u8g2, 1, 18, "U8g2 on AVR");
	u8g2_SendBuffer(&u8g2);
		
	while(1){
	}
}

**Reply by olikraus** [reply]
I think the initial case is missing, but all in all it looks good.

**Reply by Grantiscool** [reply]
Ah, U8X8_MSG_GPIO_AND_DELAY_INIT. Thanks for pointing that out.

**Reply by Grantiscool** [reply]
Hi Oli,

Unfortunately I can't get the display to do anything. It doesn't appear to be a connectivity or MCU issue. Only analogue sections such as backlight and contrast controls work as expected. I'm using an ATMega 1284P @ 8 MHz (prescaler disabled in fuse).

I'm using a particular GLCD (Winstar 12232* series) which has no Enable pin, which is presumably pulled high internally. The module has the pins CE1 and CE2. I've tried both these approaches:

-Assume CS1 = CE1 and CS2 = CE2
-Assume E = CE1 and CS1 = CE2

(the latter was based on something I read here: https://github.com/olikraus/u8g2/issues/706 )			

Please see my attached code. If I can get something simple displayed, I'll be able to learn how it all works by playing with things. Sorry in advance for any glaring stupidity! My favourite programming language is solder! lol

Thanks for any help.


-Grant

//

#include <avr/io.h>
#include <u8g2.h>
#include <util/delay.h>

#define F_CPU 1000000UL

#define DISPLAY_DATA_DIR DDRC
#define DISPLAY_DATA_PORT PORTC
#define DISPLAY_DATA_PIN0 0
#define DISPLAY_DATA_PIN1 1
#define DISPLAY_DATA_PIN2 2
#define DISPLAY_DATA_PIN3 3
#define DISPLAY_DATA_PIN4 4
#define DISPLAY_DATA_PIN5 5
#define DISPLAY_DATA_PIN6 6
#define DISPLAY_DATA_PIN7 7 

#define DISPLAY_RW_DIR DDRD 
#define DISPLAY_RW_PORT PORTD
#define DISPLAY_RW_PIN 0

#define DISPLAY_CS1_DIR DDRB
#define DISPLAY_CS1_PORT PORTB
#define DISPLAY_CS1_PIN 1

#define DISPLAY_CS2_DIR DDRB
#define DISPLAY_CS2_PORT PORTB
#define DISPLAY_CS2_PIN 2

#define DISPLAY_DC_DIR DDRB
#define DISPLAY_DC_PORT PORTB
#define DISPLAY_DC_PIN 5

#define DISPLAY_RESET_DIR DDRD
#define DISPLAY_RESET_PORT PORTD
#define DISPLAY_RESET_PIN 3


#define P_CPU_NS (1000000000UL / F_CPU)

u8g2_t u8g2;

uint8_t u8x8_avr_delay(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
	uint8_t cycles;

	switch(msg)
	{
		case U8X8_MSG_DELAY_NANO:   
			break;
		case U8X8_MSG_DELAY_100NANO:    
            
#define CALL_CYCLES 26UL
#define CALC_CYCLES 4UL
#define RETURN_CYCLES 4UL
#define CYCLES_PER_LOOP 4UL

			cycles = (100UL * arg_int) / (P_CPU_NS * CYCLES_PER_LOOP);

			if(cycles > CALL_CYCLES + RETURN_CYCLES + CALC_CYCLES) 
				break;

			__asm__ __volatile__ (
			"1: sbiw %0,1" "\n\t" // 2 cycles
			"brne 1b" : "=w" (cycles) : "0" (cycles) // 2 cycles
			);
			break;
		case U8X8_MSG_DELAY_10MICRO:    // delay arg_int * 10 micro seconds
			for(int i=0 ; i < arg_int ; i++)
				_delay_us(10);
			break;
		case U8X8_MSG_DELAY_MILLI:      // delay arg_int * 1 milli second
			for(int i=0 ; i < arg_int ; i++)
				_delay_ms(1);
			break;
		default:
			return 0;
	}
	return 1;
}


uint8_t u8x8_avr_gpio_and_delay(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
	switch(msg)
	{   
        case U8X8_MSG_GPIO_AND_DELAY_INIT: 
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN0;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN1;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN2;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN3;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN4;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN5;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN6;
			DISPLAY_DATA_DIR |= 1<<DISPLAY_DATA_PIN7;
			DISPLAY_RW_DIR |= 1<<DISPLAY_RW_PIN;
			DISPLAY_CS2_DIR |= 1<<DISPLAY_CS2_PIN;
			DISPLAY_CS1_DIR |= 1<<DISPLAY_CS1_PIN;
			DISPLAY_DC_DIR |= 1<<DISPLAY_DC_PIN;
			DISPLAY_RESET_DIR |= 1<<DISPLAY_RESET_PIN;
			break;             
		
		case U8X8_MSG_GPIO_D0:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN0);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN0);
			break;
		case U8X8_MSG_GPIO_D1:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN1);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN1);
			break;
		case U8X8_MSG_GPIO_D2:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN2);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN2);
			break;	
		case U8X8_MSG_GPIO_D3:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN3);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN3);
			break;	
		case U8X8_MSG_GPIO_D4:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN4);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN4);
			break;	
		case U8X8_MSG_GPIO_D5:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN5);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN5);
			break;	
		case U8X8_MSG_GPIO_D6:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN6);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN6);
			break;	
		case U8X8_MSG_GPIO_D7:	      
			if(arg_int)
				DISPLAY_DATA_PORT |= (1<<DISPLAY_DATA_PIN7);
			else
				DISPLAY_DATA_PORT &= ~(1<<DISPLAY_DATA_PIN7);
			break;									
		case U8X8_MSG_GPIO_CS1:        
			if(arg_int)
				DISPLAY_CS2_PORT |= (1<<DISPLAY_CS2_PIN);
			else
				DISPLAY_CS2_PORT &= ~(1<<DISPLAY_CS2_PIN);
			break;
            
        case U8X8_MSG_GPIO_CS2:
           if(arg_int)
				DISPLAY_CS1_PORT |= (1<<DISPLAY_CS1_PIN);
			else
				DISPLAY_CS1_PORT &= ~(1<<DISPLAY_CS1_PIN); 
      break;    
            
		case U8X8_MSG_GPIO_DC:        
			if(arg_int)
				DISPLAY_DC_PORT |= (1<<DISPLAY_DC_PIN);
			else
				DISPLAY_DC_PORT &= ~(1<<DISPLAY_DC_PIN);
			break;
		case U8X8_MSG_GPIO_RESET:    
			if(arg_int)
				DISPLAY_RESET_PORT |= (1<<DISPLAY_RESET_PIN);
			else
				DISPLAY_RESET_PORT &= ~(1<<DISPLAY_RESET_PIN);
			break;
		
		default:
			if (u8x8_avr_delay(u8x8, msg, arg_int, arg_ptr))	// check for any delay msgs
				return 1;
			u8x8_SetGPIOResult(u8x8, 1);      // default return value
			break;
	}
	return 1;
}

int main(void)
{
    

	u8g2_Setup_sbn1661_122x32_f(&u8g2, U8G2_R2, u8x8_byte_sed1520, u8x8_avr_gpio_and_delay);
	u8g2_InitDisplay(&u8g2);
	u8g2_SetPowerSave(&u8g2, 0);
	
	/* full buffer example, setup procedure ends in _f */
	u8g2_ClearBuffer(&u8g2);
	u8g2_SetFont(&u8g2, u8g2_font_ncenB14_tr);
	u8g2_DrawStr(&u8g2, 1, 18, "Hello");
	u8g2_SendBuffer(&u8g2);
		
	while(1){
     
	}
}

**Reply by Grantiscool** [reply]
![IMG_20210224_024050](https://user-images.githubusercontent.com/78576664/108944412-968ab280-7652-11eb-8f76-f2682ddb1ec1.jpg)

The following things made U8G2 work with the WG12232A:

- Setting the RESET pin as a high output manually via. DDR* & PORT* at the very start of main.c.
- Sending the Display On command manually (this seemed to be commented out in the src file? probably something I did!)
- U8G2's "E" assigned to WG12232's "CE1" pin, and "CS" to "CE2".
- Removed U8X8_MSG_GPIO_RESET.

Cheers for creating this pretty awesome library. I'll try not to break it!

-G

PS:- NOTE FOR FUTURE USERS: Not all WG12232* displays are created equal, with some lacking an internal clock and some having all E, CE1 and CE2 pins.

**Reply by olikraus** [reply]
:+1: 
I am happy to see the picture :-)

**Comment by olikraus** [comment]
BTW: You could still use old u8glib, it might be better optimized for AVR systems, but it is difficult for me to provide any support for u8glib.
You probably can just use the init procedure as mentioned here:
https://github.com/olikraus/u8glib/blob/f2e0c706c9462afd488023bfc81ff2d7aa4e2938/sys/atmega/hello_world/hello_world.c#L54-L55

u8glib includes a lot of support for AVR which was removed for u8g2 to make u8g2 more portable. So the advantage of u8glib would be, that you do not need to write a callback function for the IO and delay procedures.

**Reply by Grantiscool** [reply]
Hi Oli, 
Thanks SO much for taking the time. I'll have a good read and try to understand it all tonight (when my toddler stops kicking me in the brain!).

-Grant
